set.mm

```diff
@@ -4464,7 +4464,7 @@ use disjunction (although this is not required since definitions are
   $( Join antecedents with conjunction.  Theorem *3.21 of [WhiteheadRussell]
      p. 111.  (Contributed by NM, 5-Aug-1993.) $)
   pm3.21 $p |- ( ph -> ( ps -> ( ps /\ ph ) ) ) $=
-    ( wa id expcom ) BABACZFDE $.
+    ( wa pm3.22 ex ) ABBACABDE $.
```

Comment:
The new proof is shorter in the normal form. If you type `sh pr pm3.21 /norm`, the original is 39 bytes while the new one is 36 bytes. This is something the minimizer always checks, but it's rare to happen.

For you it won't say anything, but for me the minimizer says `Proof of pm3.21 stayed at 26 bytes using "ex". (Uncompressed steps decreased from 10 to 9).` (and in fact you can verify with `show new_proof /all` that the new one has ones less step).

Reply:
In https://us.metamath.org/mpeuni/conventions.html , paragraph "Acceptable shorter proofs",  you will find the rules by which one proof is considered shorter than another.  The first criterion is the length of the compressed proof, or, if that does not shrink, a second criterion is the count of symbols used for the proof in the HTML display.  Since the first criterion does not apply here, then perhaps the second?   I see exactly the same number for each version.
So what makes the new proof "shorter" then?  In NM's proof the first line is built up from an intermediate ( ph -> ps ) expression,, then id is applied to it.  In the new proof, pm3.22 gives you a complete expression ( ( ph /\ ps ) -> ( ps /\ ph ) ) right from the start, no intermediate construction step is necessary here.  I think it is disputable whether the second version (omits a non-visible construction step) is preferred to the first one (depends on more basic theorems) .  In the end, we (or I) have to review proofs, and, although having years of experience in doing so, am not able to see an advantage here.  Things get worse when proofs become longer.   I am not in favour of creating a further ternary criterion, to eliminate invisible steps existing on a technical level.

Reply:
Mh, the change is pretty minor so I guess I can revert it, but the thing is, what does the new version have that is inconvenient compared to the old one? This is what I was referring to:

Old proof of `pm3.21`:
<pre>
MM-PA> sh ne /all
 1   wph=wps       $f wff ps
 2   wps=wph       $f wff ph
 3     wph=wps       $f wff ps
 4     wps=wph       $f wff ph
 5   wch=wa        $a wff ( ps /\ ph )
 6       wph=wps       $f wff ps
 7       wps=wph       $f wff ph
 8     wph=wa        $a wff ( ps /\ ph )
 9   ex.1=id       $p |- ( ( ps /\ ph ) -> ( ps /\ ph ) )
10 pm3.21=expcom $p |- ( ph -> ( ps -> ( ps /\ ph ) ) )
</pre>

 New proof `pm3.21`:
<pre>
MM-PA> sh ne /all
1   wph=wph     $f wff ph
2   wps=wps     $f wff ps
3     wph=wps     $f wff ps
4     wps=wph     $f wff ph
5   wch=wa      $a wff ( ps /\ ph )
6     wph=wph     $f wff ph
7     wps=wps     $f wff ps
8   ex.1=pm3.22 $p |- ( ( ph /\ ps ) -> ( ps /\ ph ) )
9 pm3.21=ex   $p |- ( ph -> ( ps -> ( ps /\ ph ) ) )
</pre>

We can see that one wff step dropped.

Old proof in the normal from:
<pre>
MM-PA> sh ne /norm
Proof of "pm3.21":
---------Clip out the proof below this line to put it in the source file:
    wps wph wps wph wa wps wph wa id expcom $.
---------The proof of "pm3.21" (39 bytes) ends above this line.
</pre>

New proof in the normal from:
<pre>
MM-PA> sh ne /norm
Proof of "pm3.21":
---------Clip out the proof below this line to put it in the source file:
    wph wps wps wph wa wph wps pm3.22 ex $.
---------The proof of "pm3.21" (36 bytes) ends above this line.
</pre>

The new normal proof lists 9 labels instead of 10, making it 3 bytes shorter.

The website proof corresponds to `SHOW PROOF <label> /LEMMON /ESSENTIAL /RENUMBER /NO_REPEATED_STEPS`, so the difference is not visible there.

> I am not in favour of creating a further ternary criterion, to eliminate invisible steps existing on a technical level.

I think the main point is whether this behaviour could happen with the regular minimizer as well, or if it's just a result of my edited one. I'm pretty sure I've already seen `Proof of <label1> stayed at <number1> bytes using "<label2>".  (Uncompressed steps decreased from <number2> to <number3>)` with the regular minimizer. I guess maybe the regular minimizer never does that with wff steps?

Reply:
The minimizer comes up with suggestions.  The results here show that they are often better than the current solution.  But there are exceptions: discouraged tags are not obeyed, axioms are introduced, you listed them.  And one can add to this list:  The compressed proof size does not shrink.  In all these cases simply discard the suggestion.

The current metrics by which the the length of a proof is evaluated, is human centered.  If a proof on the HTML page appears shorter to a human, take it.  In order to have a technical means in case of contention count the symbols needed to display the proof.  As a convenient guess the length of the compressed proof turned out to also be a good indicator.

One can change the metrics to better match the results of a particular computer program instead.  But this needs a discussion in the issues first.  And you need some reasons to convince me there.

Reply:
> discouraged tags are not obeyed, axioms are introduced, you listed them. And one can add to this list: The compressed proof size does not shrink. In all these cases simply discard the suggestion.

Ok, so what you're saying makes me think that solely dropping a wff step without compressed proof shrinkage wouldn't happen with the regular minimizer, considering also the fact that nobody mentioned this issue before when I did a global minimization https://github.com/metamath/set.mm/pull/3127 (I will attempt to produce a minimal test to get some evidence of this assumption).

And even in the case this would happen with the regular minimizer it would be seen it as a bug, since this is considered an undesired change and manually editing out undesired changes is humanly inconvenient for big minimizations. Correct?

I'm going to revert pm3.21 and pm2.8 when I'm back home.

Reply:
I'm not sure whether I am fully understood.  Each theory is based on assumptions and foundations.  One of these is Metamath idea of a primitive proof step.  It presents its readers with a proof database where each proof is broken down to simple variable substitutions in a formula.  You can look at the introductory pages to verify that.  Is it primitive from a programmer's point of view?  Of course not. Under the hood, you need a grammar to dissect a formula into individual tokens, then you must employ a type checking for variable replacements, then you need a text processing unit to accomplish this all, which in turn can be broken down to individual character read, write and compare instructions.  But all this does not matter for Metamath.  Coming up with a proof that minimizes read cycles is as unimportant as one that minimizes type checking, as your version does.  This all lives under the surface.  So a minimizer should respect the foundational level of Metamath and not strive for improvements seen in a tech community only.

Reply:
I understand what you're saying, but I'm focusing on what this means in practice. I think a minimizer that produces unconventional changes is not good, because it makes the job harder for both the contributor and the reviewer. That's why I'm insisting on making sure that this case does not happen in the regular one.